### PR TITLE
Proposal to Allow Composition Functions to Set Claim Conditions

### PR DESCRIPTION
**Pull request by dalton-hill-0**
Related to #5402

### CODE REVIEW

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [inline design/one-pager-fn-claim-conditions.md:54]: Did you check that this was the case? I know we thought it was, but I hadn't verified myself.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [inline design/one-pager-fn-claim-conditions.md:167]: I might have been unclear WRT my position on events yesterday.

I don't think we need to _load_ events into the function pipeline, i.e. send composed resource events to functions in `RunFunctionRequest`. I could imagine doing that in future to give function authors more context they could use to set status conditions and/or produce results, but I would defer that until we know we need it.

I do like the idea of allowing function authors to emit "claim aligned" results (i.e. events). So something like:

```protobuf
// A Result of running a Function.
message Result {
  // Severity of this result.
  Severity severity = 1;

  // Human-readable details about the result.
  string message = 2;

  // Targets of the result. The result will be associated with each target.
  // If no targets are specified, the result targets  only the composite.
  repeated Target targets = 3;
}

// Target of Function results.
enum Target {
  TARGET_UNSPECIFIED = 0;
  TARGET_COMPOSITE = 1;
  TARGET_CLAIM = 2;
}
```

What I'm not sure about is how results should interact with status conditions.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [inline design/one-pager-fn-claim-conditions.md:16]: A little detail about what things set status conditions and emit events would be useful here. I found myself going to look it up to refresh my memory. For the record:

* A Fatal result is emitted as a warning event associated with the XR.
* A Fatal result sets the `Synced: False` status condition. The result's message is included in the condition.
* Normal and warning results emit an event associated with the XR. The reason is `ComposeResources`.
* Normal and warning results don't influence any status condition.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [reply design/one-pager-fn-claim-conditions.md:167]: Maybe there's a hybrid approach between what I sketched out in the preceding comment and what's currently in the proposal:

* Let results explicitly target an XR, claim, or both.
* Let results optionally update a status condition, in addition to emitting an event.
* If the pipeline produces >1 result targeting the same status condition, only the last one is set (the others are just events).

So something like this:

```protobuf
// A Result of running a Function.
message Result {
  // Severity of this result.
  Severity severity = 1;

  // Human-readable details about the result.
  string message = 2;

  // Machine-readable PascalCase reason for the result.
  string reason = 3;

  // Targets of the result. The result will be associated with each target.
  // If no targets are specified, the result targets only the composite.
  repeated Target targets = 4;

  // Optionally update the supplied status condition on all targets.
  // The result's reason and message will be used in the condition.
  optional Condition condition = 5;
}

// Target of Function results.
enum Target {
  TARGET_UNSPECIFIED = 0;
  TARGET_COMPOSITE = 1;
  TARGET_CLAIM = 2;
}

message Condition {
  // Type of the condition, e.g. DatabaseReady.
  // 'Ready' and 'Synced' are reserved for use by Crossplane.
  string type = 1;

  // Status of the condition.
  Status status = 2;
}

// Status of Function result condition.
enum Target {
  STATUS_UNSPECIFIED = 0;
  STATUS_FALSE = 1;
  STATUS_TRUE = 2;
}
```

File: design/one-pager-fn-claim-conditions.md
Comment (dalton-hill-0) [reply design/one-pager-fn-claim-conditions.md:54]: Yes I verified this is the case (at least in the `master` branch).

File: design/one-pager-fn-claim-conditions.md
Comment (dalton-hill-0) [reply design/one-pager-fn-claim-conditions.md:167]: Note that `Target` would specifically refer to the receiver of the `Event`. The status condition would need to be set on the XR if we want the Claim to receive them. Unless if we want to create another way to copy the conditions from the XR to the Claim, such as creating a field on the XR specifically for claim-only conditions.

Comment (dalton-hill-0) [reply design/one-pager-fn-claim-conditions.md:167]: > Let results optionally update a status condition, in addition to emitting an event.

IMO we should also enable the user to set a condition without explicitly setting an event. The user could let Crossplane decide when to emit events.

We could have logic at the reconciler that emits an event if:
`condition_is_new || condition_has_changed`

`condition_is_new` means the condition.Type was not seen as part of the previous conditions.
`condition_has_changed` means the condition.Type was seen previously, but the condition.Status has changed.

Comment (negz) [reply design/one-pager-fn-claim-conditions.md]: > Note that Target would specifically refer to the receiver of the Event. The status condition would need to be set on the XR if we want the Claim to receive them.

I'm proposing that the target should refer to the receiver of the event, _and_ of the status condition (if there is one).

So for example if a function emits a result with `TARGET_CLAIM` and a `Condition` message, I would propose that event and status condition _only_ appear on the claim, not on the XR.

> Unless if we want to create another way to copy the conditions from the XR to the Claim, such as creating a field on the XR specifically for claim-only conditions.

Right. Admittedly my proposal is focused on my preferred UX, and ignores how we'd actually build it. Of all the ways we could have a function pipeline influence claim status conditions, what you suggest seems like it would probably be the least bad.

The other alternatives that came to mind, that I like less:

* Have the XR controller mutate claims directly.
* Have the XR controller emit events about claims, and infer status conditions from those.

Comment (negz) [reply design/one-pager-fn-claim-conditions.md]: > how we'd actually build it

Maybe we eliminate `TARGET_CLAIM` only as an option? So more like:

```proto
message Result {
  // Omitted for brevity

 Target target = 4;
}

// Target of Function results.
enum Target {
  TARGET_UNSPECIFIED = 0;
  TARGET_COMPOSITE_ONLY = 1;
  TARGET_COMPOSITE_AND_CLAIM = 2;
}
```

We could then keep a mask of XR status conditions that should propagate to the claim.

My thinking is that if the point of claim-aligned status conditions and events is to produce user-friendly information, it doesn't hurt anything for that user-friendly information to appear on the XR too.

Comment (negz) [reply design/one-pager-fn-claim-conditions.md]: > IMO we should also enable the user to set a condition without explicitly setting an event. The user could let Crossplane decide when to emit events.

That would be a departure from what we do today. I believe:

* Every status condition transition produces an event (a record of the status change).
* Some things produce only events.

That's what I was going for with my proposed API:

* Every result is an event.
* Some results also change the status condition.

Comment (dalton-hill-0) [reply design/one-pager-fn-claim-conditions.md]: >That would be a departure from what we do today.

I'm just saying to provide a mechanism to turn off "Every result is an event".
If this option is turned off by the user, then "Every status condition transition produces an event (a record of the status change)." would take care of emitting the event.

By default, every result would be an event (e.g., calling `response.Fatal`). But if a user opts into a new API (e.g., `response.Condition`), then by default we would not emit the event unless the condition has changed (possibly no new code, just letting the existing condition logic take effect).

File: design/one-pager-fn-claim-conditions.md
Comment (dalton-hill-0) [inline design/one-pager-fn-claim-conditions.md]: Note: I removed this part of the example as it only works if you have a single function in the pipeline.

If you have more than one function in the pipeline and the first one returns a fatal result, the remaining functions will not be called. So pre-loading conditions like this would be limited to single function pipelines unless we change the behavior, which seems like a pretty significant change IMO.

File: design/one-pager-fn-claim-conditions.md
Comment (dalton-hill-0) [reply design/one-pager-fn-claim-conditions.md]: Good idea, I added this.

File: design/one-pager-fn-claim-conditions.md
Comment (dalton-hill-0) [reply design/one-pager-fn-claim-conditions.md]: I updated the doc to contain some of your code snippets and examples of the function author creating events and conditions that target the claim.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [inline design/one-pager-fn-claim-conditions.md:180]: FWIW I'm not sure I want to adopt the builder pattern in the Go SDK. That said, this is fine as an illustration. 🙂

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [inline design/one-pager-fn-claim-conditions.md:128]: Any thoughts on how the claim reconciler will know which status conditions it should copy from the XR?

Perhaps just a simple array of types in the XR status, e.g.:

```yaml
status:
  # The XR's condition types that should be back-propagated to the claim
  claimConditions: [DatabaseReady, ImageReady]
  # The XR's conditions
  conditions:
  - type: DatabaseReady
    status: True
    reason: Available
  - type: ImageReady
    status: False
    reason: NotFound
    message: The image provided does not exist or you are not authorized to use it.
```

**Review by negz — COMMENTED** [review]
I'm liking how this is looking, thanks @dalton-hill-0!

CC @phisco for a second opinion from the maintainer team.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [reply design/one-pager-fn-claim-conditions.md]: > I am just thinking that in the case of results containing conditions, it may be nice to reduce event count if conditions are not changing.

Yeah good call - we don't want to emit an event if it's effectively just capturing that the status is unchanged.

File: design/one-pager-fn-claim-conditions.md
Comment (dalton-hill-0) [reply design/one-pager-fn-claim-conditions.md:180]: > FWIW I'm not sure I want to adopt the builder pattern in the Go SDK

I'm not sold on it either; I'm open to ideas.
The idea behind the builder was due to the Result being more complex at this point and the handling of args was beginning to get a little messy IMO.

Here are some thoughts on the subject:
```go
// expose all args. Not great as condition and target (should?) likely be optional
response.Normal(rsp, message, target, conditionType, conditionStatus, conditionReason)

// we could make both target and condition pointers, handling for nil
condition := Condition(type, status, reason)
response.Normal(rsp, message, target, condition)

// Perhaps an opts approach?
response.Normal(rsp, message, ...opts)

// We could supply predefined opts like Condition(type, status, reason) and Target_X_()
response.Normal(rsp, message, TargetClaim(), Condition("DatabaseReady", false, "Unauthorized"))
```

File: design/one-pager-fn-claim-conditions.md
Comment (dalton-hill-0) [reply design/one-pager-fn-claim-conditions.md:128]: If we are fine adding a new field to status, can we just have `claimConditions` hold the actual conditions themselves? This would allow Result.Target to choose Claim directly, instead of CompositeAndClaim.

One benefit of this would be internal errors.
Claim would see:
```yaml
  - type: DatabaseReady
    status: False
    reason: InternalError
    message: "An internal error occurred. If you need immediate assistance, please contact us at _XYZ_."
```

Composite would see:
```yaml
  - type: DatabaseReady
    status: False
    reason: InternalError
    message: "Failed to parse response from AWS: _some_error_"
```

**Comment by negz** [comment]
Out of scope for this design, but I could imagine building a generic function to expose this functionality to folks. Here's a rough sketch.

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: function-template-go
spec:
  compositeTypeRef:
    apiVersion: example.crossplane.io/v1
    kind: NoSQL
  mode: Pipeline
  pipeline:
  - step: patch-and-transform
    functionRef:
      name: function-patch-and-transform
    input:
    # Omitted for brevity
  - step: claim-status
    functionRef:
      name: function-result-extractor
    input:
      apiVersion: extractor.fn.crossplane.io/v1alpha1
      kind: Results
      results:
      - result:
          target: Claim
          severity: Warning
          message: Internal database error: ${error}
          status:
            type: DatabaseReady
            condition: "False"
            reason: InternalError
        triggers:
        - resourceName: rdsinstance
          type: StatusCondition
          status:
            type: Synced
            condition: "False"
            message: "Some AWS error: (?P<error>)"
```

The idea is you configure an array of results. Each result takes a series of triggers. If all triggers are true, the result is emitted. Each trigger uses regular expressions to match status conditions of composed resources.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [reply design/one-pager-fn-claim-conditions.md:180]: I like the variadic functions as arguments approach you've got sketched out in the last alternative best. The only gotcha would be `response.Normalf` etc, which already accept variadic format args.

I think there's some precedent for a slightly lighter weight version of the builder pattern in crossplane runtime, e.g. we have:

```go
c := xpv1.Available().WithMessage("foo")
```

Where `WithMessage` returns the underlying condition with the message set.

So I could probably live with:

```go
response.Normalf(rsp, message, args...).
    TargetClaim().
    Condition("DatabaseReady, false, "Unauthorized")
```

**Comment by dalton-hill-0** [comment]
I added a WIP PR for the Crossplane changes (#5450).
Tests are not completed as the design is not yet agreed on.

**Comment by negz** [comment]
@dalton-hill-0 Sorry, I'm running even more behind on design reviews than I expected. I'll hopefully have time to focus on this this week though.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [reply design/one-pager-fn-claim-conditions.md:128]: Quickly capturing my take from the last functions SIG meeting: I prefer the "mask" approach where we store a set of condition types that also apply to the claim.

The rationale:

* The claim is a user-facing presentation of the XR, so all claim conditions should also apply to the XR.
* Two arrays of conditions on the XR seems easier to confuse - which are true for the XR?

**Comment by dalton-hill-0** [comment]
@negz - FYI I just updated the [PR](https://github.com/crossplane/crossplane/pull/5450) to set `ClaimConditions` to a list of strings. Additionally, it contains a generic way of setting conditions that will emit events only if the condition has changed.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [reply design/one-pager-fn-claim-conditions.md:128]: Can you add this to the design doc somewhere too, please.

File: design/one-pager-fn-claim-conditions.md
Comment (negz) [reply design/one-pager-fn-claim-conditions.md:180]: @dalton-hill-0 Could you update the design doc to make it clear that the builder pattern example is just a sketch, not necessarily the API we're committing to.

**Comment by negz** [comment]
@dalton-hill-0 I just added a comment to each of the two open comment threads. If you can address those, I'm happy to merge this design and we can focus on the implementation PR. Thanks!

**Comment by dalton-hill-0** [comment]
> @dalton-hill-0 I just added a comment to each of the two open comment threads. If you can address those, I'm happy to merge this design and we can focus on the implementation PR. Thanks!

@negz - these have been addressed

**Review by negz — APPROVED** [review]
Awesome, thank you @dalton-hill-0!

**Comment by raphasle** [comment]
I stumbled upon this PR after realizing that functions can't write to the spec of the composite resource (as documented [here](https://docs.crossplane.io/v1.16/concepts/composition-functions/#desired-state)). I used this functionality before using pipeline mode to write default values back to the claim (e.g. spec.version for a Postgres DB or spec.location for an Azure Storage Account) if the developer did not specify this optional parameter in the claim.
I know this is a bit off-topic here but maybe one of you can point me to a workaround example showing how to achieve this or to another design document saying why it should not be possible anymore to set fields in the spec (and metadata) section of the composite resource?

**Comment by negz** [comment]
@raphasle The short version is that it's hard to avoid bugs when we're syncing claim <-> composite <-> composed resource. See https://github.com/crossplane/crossplane/issues/4581.

**Pull request merged**
